### ISSUE TITLE
Slow/blocking client initialization with many persistent messages

### ISSUE THREAD

**Issue by mecraggs** [eclipse-paho/paho.mqtt.c#835]
**Describe the bug**
When using the async client with default persistence and automatic reconnect the number of buffered messages is reset although there are buffered messages stored on disk and then fills up all available disk space. 
The client at no point has a connection to the broker. 
Using paho as submodule from tag 1.3.1.  

**To Reproduce**
```
 MQTTAsync_createOptions opts = MQTTAsync_createOptions_initializer5;
    opts.maxBufferedMessages = p_numPersistence;
    opts.sendWhileDisconnected = 1;
    const std::string broker =
        m_brokerAddress + ":" + std::to_string(m_brokerPort);
    m_initialized = MQTTAsync_createWithOptions(
        &m_client, broker.c_str(), m_clientId.c_str(),
        MQTTCLIENT_PERSISTENCE_DEFAULT, const_cast<char*>(PERSISTENT_STORAGE),
        &opts);
```

```
  MQTTAsync_connectOptions opts = MQTTAsync_connectOptions_initializer;
    opts.username = m_user.c_str();
    opts.password = m_password.c_str();
    opts.automaticReconnect = 1;
    opts.cleansession = 1;
    opts.connectTimeout = 1;
    opts.context = m_client;
    opts.onFailure = onConnectFailure;
    opts.onSuccess = onConnect;
```

**Environment**
 - OS: Linux
 - Version 1.3.1

**Additional context**
I've added some debug output, which reproduces the following log output. 
```
#include <stdio.h>
int MQTTAsync_send(MQTTAsync handle, const char *destinationName,
                   int payloadlen, const void *payload, int qos, int retained,
                   MQTTAsync_responseOptions *response) {
  int rc = MQTTASYNC_SUCCESS;
  MQTTAsyncs *m = handle;
  MQTTAsync_queuedCommand *pub;
  int msgid = 0;

  printf("Buffered messages: %d | Max: %d\n",
         MQTTAsync_countBufferedMessages(m),
         m->createOptions->maxBufferedMessages);
```
```
Buffered messages: 3278 | Max: 86400
Buffered messages: 3279 | Max: 86400
Buffered messages: 3280 | Max: 86400
Buffered messages: 3281 | Max: 86400
Buffered messages: 3282 | Max: 86400
Buffered messages: 3283 | Max: 86400
2020-03-27 14:54:17,817686 Detections: 6
Buffered messages: 3284 | Max: 86400
Buffered messages: 3285 | Max: 86400
2020-03-27 14:54:18,218691 Mqtt failed to connect!
Buffered messages: 1 | Max: 86400
Buffered messages: 2 | Max: 86400
Buffered messages: 3 | Max: 86400
Buffered messages: 4 | Max: 86400
```
* Am I missing a flag somewhere ? 
* I did not have this issue when using the cpp client ( we moved away because we don't want to use exceptions ) 
* Any pointers would be greatly appreciated :)

**Comment by icraggs** [comment]
As the C++ client uses this library underneath, the behaviour should be the same, obviously!

What QoS are you using?  Given that QoS 0 is 'fire and forget' there have always been differing opinions on whether QoS 0 messages should be deleted when the MQTT session state is cleared.

**Comment by mecraggs** [comment]
I am using qos 1.
Absolutely agree that the behavior should be the same. 
I will take some time debug into it.

**Comment by mecraggs** [comment]
Sorry for the delay, we had other issues to sort out, but I believe to have found the problem, 
During porting the property 'clean session' was set wrongly set to true, which would explain the behavior. With set to false, it works as expected. Apologies for the noise. 

Before closing this ticket, I would have another question: 
When starting up and there are a lot of messages persisted on disk it takes a very long time for the mqtt client to start ( not connecting just creating the client ). We run in a limited environment with 512 to 1024 mb of memory and with 50k messages on disk the container is killed when creating the mqtt client. 
Is there a way to mitigate this other than reducing the number of maximum buffered messages or creating a custom persistence store ?

**Comment by icraggs** [comment]
No problem with the delay.  I plan to be focussing on persistence in the 1.3.3 release, so you could change the title of this issue so we could think about it.  

I would say that:

1. the persistence wasn't implemented with the idea that as many as 50k messages would be stored in it.  
2. I don't think we should reimplement database functionality.

Of course we could add a more lazy reading of data from the store, and/or provide a link to a proper database as options off the top of my head.

**Comment by icraggs** [comment]
I've added some changes to the persistence loading for messages stored on disk.

On my Linux desktop, this reduces loading times from 28s to 1.4s for 64000 messages.  On my macbook, from 164s to 4s.

(I hadn't originally intended so many messages to be stored, so I didn't really think about it.  I replaced a linear search when inserting messages in order with a binary one.)

**Comment by mecraggs** [comment]
Hi, 
Sounds great, really appreciate it. I hope to give it a try on our side sometime this week. 
I assume this will be included in the 1.3.6 release ?

**Comment by icraggs** [comment]
> Hi,
> Sounds great, really appreciate it. I hope to give it a try on our side sometime this week.
> I assume this will be included in the 1.3.6 release ?

Yes
